confgenerator/jvm.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package confgenerator
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/otel"
+	"github.com/kardianos/osext"
+)
+
+type MetricsReceiverJVM struct {
+	ConfigComponent `yaml:",inline"`
+
+	MetricsReceiverShared `yaml:",inline"`
+
+	Endpoint string `yaml:"endpoint" validate:"omitempty,url"`
+}
+
+const defaultJVMEndpoint = "localhost:9999"
+
+func (r MetricsReceiverJVM) Type() string {
+	return "jvm"
+}
+
+func (r MetricsReceiverJVM) Pipelines() []otel.Pipeline {
+	if r.Endpoint == "" {
+		r.Endpoint = defaultJVMEndpoint
+	}
+
+	jarPath, err := findJarPath()
+	if err != nil {
+		_ = err // TODO(djaglowski) return nil, fmt.Errorf("create jvmreceiver: %v, err)
```

Comment:
Is `return nil, fmt.Errorf("create jvmreceiver: %v, err)` commented out here by accident or on purpose?

Reply:
This is on purpose, but we should discuss.

How would you like errors to be handled in this method? 

Should `Pipelines()` return `([]otel.Pipeline, error)`?

Reply:
Sync'd with Quentin offline.

**TL;DR**

The suggestion is to embed the jar in the binary for now (https://pkg.go.dev/embed) to bypass the need to throw this error and get unblocked quickly. This buys us some time to come up with a cleaner solution for similar needs (e.g. JAVA_HOME) as elaborated below.

**The long story**

In general, user config errors should fail config generation entirely instead of leaving the agent partially working, so that users can quickly detect that something is wrong and fix them. This helps limit the number of customer cases when the root cause is user config error.

However, the error here is about a case when we can't find the jar or Java etc., which is a mix of user error and missing an external dependency.

If in the future we have better diagnose signals (e.g. uptime metrics for each receivers), it's more reasonable to let things fail partially. Today if we let the agent work partially, the service status will still show it as normal. Unless the user looks at the agent logs, or try to query JVM metrics, they won't notice that something is broken, which could be a confusing UX. So for now we'd like to stick with failing the whole config generation.

That means we need a way to return errors from receivers. For configurable paths (e.g. the java home), it could be done through field validation. For things that are not user-configurable, we could allow receivers to define a Validate method to perform extra validation, and if it exists, we'll arrange for that to be called as part of overall validation.

Another idea is to build a mechanism to compute field defaults, and then rely on the existing validation logic to fail if those fields are blank. For example, `JAVA_HOME` should have a computed default, that the user can override, and then only if both of those fail and the field is still empty, we return a config error.